client_test.go

```diff
@@ -591,22 +591,98 @@ func TestClientTopics(t *testing.T) {
 }
 
 func TestClientSetIDKey(t *testing.T) {
-	clientID := e4crypto.HashIDAlias("client1")
-	validKey := e4crypto.RandomKey()
-	invalidKey := make([]byte, e4crypto.KeyLen)
+	_, edSk, err := ed25519.GenerateKey(nil)
+	if err != nil {
+		t.Fatalf("failed to generate ed25519 key")
+	}
 
-	c, err := NewClient(&SymIDAndKey{
-		ID:  clientID,
-		Key: validKey,
-	}, NewInMemoryStore(nil))
+	c2PubKey := make([]byte, e4crypto.Curve25519PubKeyLen)
+	_, err = rand.Read(c2PubKey)
 	if err != nil {
-		t.Fatalf("Failed to create client: %v", err)
+		t.Fatalf("failed to generate c2 pubkey: %v", err)
+	}
+
+	symConfig := &SymIDAndKey{
+		ID:  e4crypto.HashIDAlias("client1"),
+		Key: e4crypto.RandomKey(),
 	}
 
-	if err := c.setIDKey(invalidKey); err == nil {
-		t.Fatal("Expected an error when calling setIDKey with an invalid key")
+	pubConfig := &PubIDAndKey{
+		ID:       e4crypto.HashIDAlias("client1"),
+		Key:      edSk,
+		C2PubKey: c2PubKey,
 	}
 
+	t.Run("Setting a valid symkey to a symkey client", func(t *testing.T) {
+		c, err := NewClient(symConfig, NewInMemoryStore(nil))
+		if err != nil {
+			t.Fatalf("Failed to create client: %v", err)
+		}
+
+		validSymKey := e4crypto.RandomKey()
+		if err := c.setIDKey(validSymKey); err != nil {
+			t.Fatalf("failed to set valid symkey: %v", err)
+		}
+	})
+
+	t.Run("Setting a invalid symkey to a symkey client", func(t *testing.T) {
+		invalidKeys := [][]byte{
+			make([]byte, e4crypto.KeyLen),
+			e4crypto.RandomKey()[:e4crypto.KeyLen-1],
+			edSk,
+		}
+
+		c, err := NewClient(symConfig, NewInMemoryStore(nil))
+		if err != nil {
+			t.Fatalf("Failed to create client: %v", err)
+		}
+
+		for _, key := range invalidKeys {
+			if err := c.setIDKey(key); err == nil {
+				t.Fatal("an error was expected when setting an invalid symkey")
+			}
+		}
+	})
+
+	t.Run("Setting a valid ed25519 key to a pubkey client", func(t *testing.T) {
+		c, err := NewClient(pubConfig, NewInMemoryStore(nil))
+		if err != nil {
+			t.Fatalf("Failed to create client: %v", err)
+		}
+
+		_, validEdSk, err := ed25519.GenerateKey(nil)
+		if err != nil {
+			t.Fatalf("failed to generate ed25519 key: %v", err)
+		}
+
+		if err := c.setIDKey(validEdSk); err != nil {
+			t.Fatalf("failed to set valid ed25519 key: %v", err)
+		}
+	})
+
+	t.Run("Setting an invalid ed25519 key to a pubkey client", func(t *testing.T) {
+		_, validEdSk, err := ed25519.GenerateKey(nil)
+		if err != nil {
+			t.Fatalf("failed to generate ed25519 key: %v", err)
+		}
+
+		invalidKeys := [][]byte{
+			make([]byte, ed25519.PrivateKeySize),
+			validEdSk[:ed25519.PrivateKeySize-1],
+			e4crypto.RandomKey(),
+		}
+
+		c, err := NewClient(pubConfig, NewInMemoryStore(nil))
+		if err != nil {
+			t.Fatalf("Failed to create client: %v", err)
+		}
+
+		for _, key := range invalidKeys {
+			if err := c.setIDKey(key); err == nil {
+				t.Fatal("an error was expected when setting an invalid ed25519 private key")
+			}
+		}
+	})
 }
```

Comment:
Notice that you are iterating over keys and then invoking t.Fatal on the first failing one? That's a sign that these sub-tests really should be tests of their own, they should be broken out.

```go
func TestClientSetIDKey_invalidEd25519Key(t *testing.T) {
                 client, err := NewClient(pubConfig, NewInMemoryStore(nil))
		if err != nil {
			t.Fatalf("Failed to create client: %v", err)
		}

		invalidKeys := []struct{
                      name string
                      key []byte
                 } {
			{"all zeros", make([]byte, ed25519.PrivateKeySize)},
			{"ed25519.PrivateKeySize-1", validEdSk[:ed25519.PrivateKeySize-1]},
			{"random key", e4crypto.RandomKey()},
		}
		for _, tt := range invalidKeys {
                         t.Run(tt.name, func(t *testing.T) {
			        if err := client.setIDKey(tt.key); err == nil {
				            t.Fatal("expected an error")
			       }
                          })
		}
}
```

and do the same for all the other tests.